### PR TITLE
Rework many POSIX and C lib with new trait feature

### PR DESCRIPTION
**Pull request by ekrich**
Just trying to do little heavy lifting to get towards a POSIX lib that can be used on its own.

There are definitely cases where it is unclear where things should go especially for types. In C we can have header files defined once so they can be referenced more than one place and they won't conflict. In Scala we could have them as `type`s in more than one place and they could conflict on `import`. This is an area we need to work out best practices.

This still doesn't solve the "I have to look at 2 specs at the same time" problem but at least we know it can be defined once either in C shared with POSIX or POSIX only.

In general in the `javalib` with could see `libc` imports because lib C can do some things but it may be more likely that we see `posix` or `windows` as the main imports for native code.

For a developer, they should be able to just import `libc` or `posix` depending on what they need. We still have quite a bit of work in this area but it is getting better.

### CODE REVIEW

File: clib/src/main/scala/scala/scalanative/libc/string.scala [hunk 1]
```diff
@@ -3,8 +3,10 @@ package libc
 
 import scalanative.unsafe._
 
-@extern
-object string {
+@extern object string extends string
+
+/** Currently following C99 */
```
Comment (LeeTibbert) [inline clib/src/main/scala/scala/scalanative/libc/string.scala:8]: Better with the comment, but left wondering...

Could the Scaladoc comment be extended by a sentence
or two.

Why & how are these C99?   We compile using C11.
Certainly do not want to break users who compile with C99
but it seems like an odd restriction.

File: posixlib/src/main/scala/scala/scalanative/posix/string.scala [hunk 2]
```diff
@@ -14,9 +15,9 @@ import scala.scalanative.posix.sys.types
  *  A method with an XSI comment indicates it is defined in extended POSIX
  *  X/Open System Interfaces, not base POSIX.
  */
+@extern object string extends string
 
-@extern
-object string {
+@extern trait string extends libc.string {
```
Comment (LeeTibbert) [inline posixlib/src/main/scala/scala/scalanative/posix/string.scala:20]: The "CX" Scaladoc convention is probably obsolete/unnecessary now.  I think it tells users
nothing and imposes a cost on maintainers.

Besides, the C that CX is an extension of is IEEE C,
not the ISO C of clib. So, we are pointing out an 
extension to something we do not directly support/implement.

XSI, PS, & kin are probably still useful.

**Comment by ekrich** [comment]
@LeeTibbert I can address the issues above. If you have a minute to look and see what might be wrong with the Windows builds, I would be appreciative.

File: clib/src/main/scala/scala/scalanative/libc/math.scala [hunk 3]
```diff
@@ -3,12 +3,15 @@ package libc
...
 
-@extern
-object math {
+@extern object math extends math
+
+@extern trait math {
 
   // Basic operations
 
+  // Not found in POSIX spec
```
[the remark this answers is not in the document]
Comment (ekrich) [reply clib/src/main/scala/scala/scalanative/libc/math.scala:12]: Ok, it was getting late so now this is fixed.

File: clib/src/main/scala/scala/scalanative/libc/string.scala (hunk 1 above)
Comment (ekrich) [reply clib/src/main/scala/scala/scalanative/libc/string.scala:8]: Removed and put the C99 reference in the docs. We can follow C11 and add additional functions if we want but we would have to decide as a team.

File: posixlib/src/main/scala/scala/scalanative/posix/inttypes.scala [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
 
-import scalanative.unsigned.{UByte, UInt, ULong, UShort}
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned.{UByte, UInt, ULong, UShort}
 
-object inttypes {
+@extern object inttypes extends inttypes
+
+@extern trait inttypes extends libc.inttypes {
+  // should be in stdint, optional in C stdint
```
[the remark this answers is not in the document]
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/posix/inttypes.scala:10]: Ok, I put all my thoughts there - they are buried but better than nothing.

File: posixlib/src/main/scala/scala/scalanative/posix/string.scala (hunk 2 above)
Comment (LeeTibbert) [reply posixlib/src/main/scala/scala/scalanative/posix/string.scala:20]: You are right. The CX is ISO C.  So IEEE C is extending ISO.    
Given the current set of definitions, why would anything ever be in a posixlib file if
it were not extending or modifying an ISO C declaration?

If everything is marked CX, then, effectively, nothing is marked CX. Simpler
to do the latter.  Discussing, not objecting.  Trying to think of what adds value
and what is simply an additional hoop for developers.

**Comment by LeeTibbert** [comment]
Re: Windows builds.   On the case.

File: posixlib/src/main/scala/scala/scalanative/posix/string.scala (hunk 2 above)
Comment (ekrich) [reply posixlib/src/main/scala/scala/scalanative/posix/string.scala:20]: Agreed, extending ISO C is certainly the plan. Without good package docs, less is probably better as you say. If they are extensions to POSIX then maybe a comment is good.

**Comment by ekrich** [comment]
@LeeTibbert It seems to only happen on Windows - I changed `libc.string` to use the new `trait` feature which is my guess because that is where `memset` is located other than in `nativelib/src/main/scala/scala/scalanative/runtime/libc.scala` which I didn't change. 

Maybe @WojciechMazur needs to take a look. It could have gotten more strict so maybe a call site is bad - could only `DWord` stuff on the Windows side.

**Comment by LeeTibbert** [comment]
It happens on Linux SN 2.11 (sometimes).  I just double checked the logs above.

I may have a clue.
```val privilegeSetLength = stackalloc[windows.DWord]()
          !privilegeSetLength = emptyPriviligesSize.toUInt

          val privilegeSet: Ptr[Byte] = stackalloc[Byte](!privilegeSetLength)
          memset(privilegeSet, 0, !privilegeSetLength)
```
The stackalloc is using a windows.DWord (as you mentioned).  Has that line
changed in this PR?   A DWord is a fixed 32 bit size, not float 32/64 for actual
architecture.  The memset declarations use CSize as the third argument. I believe
that transforms to USize, which is 64 bits on 64 bit systems.   

I did not see a Dfoo which would chameleon 32/64 bit the way that CSize will.

As a debugging shot in the dark, you could hard code that `stackalloc` to 64 bits
and see what happens.

If I am onto something, Scala 2.11 may be doing us a favor here by complaining,
in its obscure way.   The values being put into the `stackalloc`ed location are
unlikely to be larger than 32 bits, but there definitely is a theoretical buffer overrun bug there.

**Comment by WojciechMazur** [comment]
The errors are probably caused by wrong type signature by using boxed type (CSize) instead of primtive int/long. I'll try to fix this issue in the compiler plugin soon

**Review by WojciechMazur — APPROVED** [review]

**Pull request merged**